rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1272 @@
+          baz;
+      }; # <-- The decrease by two levels here is okay
+    in
+    null
+    ```
+
+### Expansion of expressions
+
+Unless stated otherwise, any expression that fits onto one single line must be trivially formatted as such.
+
+For list elements, attributes, and function arguments, the following applies:
+
+- If expanded into multiple lines, each item must be on its own line.
+  - Grouping similar items together can be done by adding blank lines or comments between the groups instead.
+  - This also applies to the first item, so e.g. `[ firstElement` in a multi line list is not allowed.
+- Long sequences of items should be liberally expanded, even if they would fit onto one line character-wise.
+  - The motivation is to keep the information per line manageable. Usually "number of elements" is a better metric for that than "line length".
+  - The cutoff is usually determined empirically based on common usage patterns.
+
+**Examples:**
+
+```nix
+{
+  buildInputs = [
+    foo
+    bar
+    baz
+
+    somethingElse
+  ];
+
+  systemd.services = {
+    foo = { };
+    bar = { };
+  };
+
+  inherit
+    lib
+    foo
+    bar
+    baz
+    ;
+}
+```
+
+### Function application
+
+- In a function application chain, the first element is treated as the "function" and the remaining ones as "arguments".
+- As many arguments as possible must be fit onto the first line.
+  - If all but the last argument do fit, then the last argument may also start on the first line.
+  - If an earlier argument does not fit onto the first line, then that argument and all the following ones must start on their own line.
+  - All arguments that are not on the same line as the function must be indented by one level.
+
+**Examples:**
+
+```nix
+# All arguments fit onto the first line
+function arg1 arg2
+
+# The line length limit is reached, so the remaining arguments need to be on their own lines
+function arg1 arg2 arg3
+  arg4
+  arg5
+
+# The last argument is a multiline expression, so it doesn't fit on the first line,
+# but it can still start on the first line
+function arg1 arg2 {
+  more = "things";
+}
+
+# The second argument doesn't fit on the first line, but it's not the last argument,
+# so it needs to start on a new line
+function arg1 arg2
+  {
+    more = "things";
+  }
+  arg3
+
+# Same with more multiline arguments
+function
+  {
+    a = 1;
+    b = 2;
+  }
+  {
+    c = 1;
+    d = 2;
+  }
+```
+
+**Drawbacks**
+
+- This style sometimes forces lists or attribute sets to start on a new line, with additional indentation of their items.
+
+**Alternatives**
+
+- Compacting multiline arguments like this:
+  ```nix
+  function arg1 {
+    # stuff
+  } arg3
+
+  function {
+    # …
+  } {
+    # …
+  }
+  ```
+  - This violates the guideline of the indentation representing the expression structure, and thus reduces readability.
+
+### Function declaration
+
+- The body of the function must not be indented relative to its first arguments.
```

Comment:
I'd like to suggest that, additionally, a first level of indentation should not be mandatory when a function declaration is wrapped in a set of parentheses to be passed as an argument to another function.

Consider this common pattern in NixOS tests:

```nix
import ./make-test-python.nix (
{ pkgs, lib, ... }:

let
  # ...
in

{
  # ...
}
)
```

Which would be reformatted as:

```nix
import ./make-test-python.nix (
  { pkgs, lib, ... }:

  let
    # ...
  in

  {
    # ...
  }
)
```

The function definition is effectively at the top level, so I think it's less readable to have the whole rest of the contents of the expression indented like this.

Reply:
Essentially, for me, it's important for readability that boilerplate doesn't introduce the first level of indentation.  Indentation is a signal to me that I should pay attention, and the point of boilerplate helpers like make-test-python.nix is that I don't need to pay too much attention to them.  The only reason they're required at all is to make files directly buildable from the Nix CLI without having to be imported/callPackaged/whatever.  It's a quirk of the language that these are function calls at all.

Reply:
We discussed this in the meeting and don't think that we should change the formatting for this. The inconsistency it introduces doesn't seem worth the reduction of one level of indentation. Our suggestion is to refactor the code to not need the boilerplate in the first place, e.g. by creating a `default.nix` that calls `make-test-python.nix` instead. And that's also what Nixpkgs' NixOS tests now do, allowing you to also call the tests with `nix-build -A nixosTests.<name>`.